### DISCUSSION TITLE
Robocode Tank Royale on macOS 12.3.1 (x64/ARM) no bot can join and error with .NET6

### DISCUSSION THREAD

**Discussion by CodeMonkeyBoy** [robocode-dev/tank-royale#20] [Q&A] [answered]
Hey,

first I want to thank you for Robocode, I like your work very much. Keep up the awesome work!

I played the "old" version a lot and saw there is a new version which I want to give it a try.
But sadly I cant't get it to work on both of my machines running macOS 12.3.1

First, it is not compatible with .NET6 LTS SDK (.NET5 will EoL May 10th 2022) they changed something in the System.Globalization so the Demo bots throw `System.Globalization.CultureNotFoundException: Culture is not supported. (Parameter 'culture') 4096 (0x1000) is an invalid culture identifier.` so it is unplayable on Apple Silicon chips because .NET5 doesn't really support these chips.
To avoid this Exception you can clear the value of `countryCodes` from the *.json in the bot folder.

Secondly it will still not work, the GUI starts, Server is up and running, but none of the bots can connect.
Bootup in the GUI seem to work but none can join, nothing shows up.

In the Terminal I tested it with the C# bots and get `Connected to ws://localhost:7654/
Disconnected from ws://localhost:7654/, status code: 1008, reason: Wrong secret` on both Machines (x64 and ARM).
So I startup the Java bots with the *.sh, same error message...wrong secret. The server writes some secret into the properties, copy that into the json as "secret" still doesn't work.

Did I miss something?
How should i configure the server to allow any bot? The GUI can connect and test the server.

Thanks in advance and best regards
Ben

**Comment by flemming-n-larsen** [comment]
Hi @CodeMonkeyBoy 

Upgrading to .Net 6 seems like a good idea, so I will look into that.

You are right that you need the secret from the `server.properties` (from `bots-secret` field).
Read more info [here](https://robocode-dev.github.io/tank-royale/articles/debug.html#run-your-bot-from-the-command-line)

The secret is not meant to be set in the bot's metadata (JSON) file, but instead as environment variable:
```bash
export SERVER_SECRET=s0m3R0bOc0dEs3crEt
```

I hope setting the SERVER_SECRET will work for you?

**Reply by CodeMonkeyBoy** [reply]
Awesome work and thanks for the Update to .NET6, I don't know how i missed that part with the Server Secret, but honestly I didn't look into the Debugging Section.
With `export SERVER_SECRET` it worked like a charm.

I saw it in the Schema of the JSON, so i thought it would go there. Nevertheless so it worked.

**Comment by flemming-n-larsen** [comment, marked as answer]
@CodeMonkeyBoy 

I have now upgraded the Bot API for .Net to use .Net 6 with [release 0.13.0](https://github.com/robocode-dev/tank-royale/releases/tag/v0.13.0)

**Reply by CodeMonkeyBoy** [reply]
Awesome, i think it will be give a boost to the future, unfortunately it sill doesn't work if countryCodes has values.
But as I mentioned above if you remove them from now it will work. I am not sure but a hint should be the the country codes in the json are to letter codes. If I remember correctly .NET 6 will handle them slightly different (e.g. us -> en-US) but if you change that the BOT API NuGet throws an exception because the letter code should not loner than two letters.

And I still can't use the Booter in the GUI, they seem to boot up but the can't be selected for join any battle.

Keep up this great work! And again, thanks for your time and support.

**Reply by flemming-n-larsen** [reply]
Sorry, version 0.13.0 did not include the other fixes, but now I will definitely have a look into:

- CountryCodes issue.
- Booting up the .Net based bots on macOS with the GUI.

**Reply by CodeMonkeyBoy** [reply]
Hey no need to apologize :)
I just want to share my experiences. With the workarounds mentioned above it is playable, and… I love it!

Thank you for have a deeper look into this issues for such a small audience. You rock! :)

**Reply by flemming-n-larsen** [reply]
The issue with the country codes failing with .Net under macOS has now been fixed (debugging and testing the issue on a MacBook with .Net 6 SDK installed).
The same issue was causing the boot up problem, so both problems is now solved, and the fix will be available with the next release.

**Reply by CodeMonkeyBoy** [reply]
Great to hear that, I‘m really excited. Thanks again :)

**Reply by flemming-n-larsen** [reply]
Fix is available with [release 0.13.1](https://github.com/robocode-dev/tank-royale/releases/tag/v0.13.1)

**Reply by CodeMonkeyBoy** [reply]
T(h)anks again! :)

Works very well!

But found another bug, booted Tanks (within GUI with Booter) won't show up as available (Joined Bots) when the path to the bot had whitespace in it. If you remove the whitespace, all work as intended.
<img width="1265" alt="joined_bots" src="https://user-images.githubusercontent.com/61388885/169118762-2b13a64c-92b3-41e0-a798-4e0c8253ef72.png">

**Reply by CodeMonkeyBoy** [reply]
Another interesting thing, if they had whitespace in their path, you can add them multiple times with "Boot ->" but they won't show up in Joined Bots. If there are no whitespace in path, you can click multiple times on "Boot ->" they shown up there once but multiple times in Joined Bots, which worked well.
<img width="1265" alt="Screen Shot 2022-05-18 at 20 43 16" src="https://user-images.githubusercontent.com/61388885/169123439-b77d159d-26ae-4ee7-a340-213dbd7bdbf8.png">

Hmm can't reproduce the second sight...now they show up multiple times in Booted aswell, but the other one:
<img width="1265" alt="Screen Shot 2022-05-18 at 20 46 37" src="https://user-images.githubusercontent.com/61388885/169125616-479dabb0-3fc2-4f75-992c-6264a1e6233a.png">

**Reply by flemming-n-larsen** [reply]
@CodeMonkeyBoy

T(h)anks for reporting the issue with whitespaces in their path. 🙂👍
I guess it is a simple issue with the booter that needs a fix.

**Reply by CodeMonkeyBoy** [reply]
You are welcome!

**Reply by flemming-n-larsen** [reply]
> @CodeMonkeyBoy
> 
> T(h)anks for reporting the issue with whitespaces in their path. 🙂👍 I guess it is a simple issue with the booter that needs a fix.

Hmm... it seems to work well, at least under Windows 11 with blank spaces in the path name. But perhaps this is an issue seen on macOS?

**Reply by CodeMonkeyBoy** [reply]
If it can't be reproduced on Windows it might be an issue only on macOS (maybe Linux too?). I can reproduce it (seen on last screenshot Corners had whitespace in path, Crazy not), it happens with both versions of the Bots (Java and C#).

Is there any way to log what happens on my machine to help you with?

**Reply by flemming-n-larsen** [reply]
I will try this out on macOS and Linux. Don't worry. 🙂

**Reply by flemming-n-larsen** [reply]
I am able to reproduce the issue on macOS. But it seems to be a tricky/nasty issue to fix. So I will look deeper into solving this.

**Reply by flemming-n-larsen** [reply]
I found the bug, which was with running a process using `bash` with the `-c` (command) option, that needed quote marks when white spaces are being used. The fix will be available with the next release. 🙂

**Reply by flemming-n-larsen** [reply]
@CodeMonkeyBoy 
The fix for whitespaces is now available with release [0.13.3](https://github.com/robocode-dev/tank-royale/releases/tag/v0.13.3) 😊

**Reply by CodeMonkeyBoy** [reply]
T(h)ank you very much!
Awesome product, keep up this great work and your support. Love it
